pkg/gardenlet/controller/managedseed/managedseed_reconciler.go

```diff
@@ -0,0 +1,280 @@
+	}
+
+	managedSeed := &seedmanagementv1alpha1.ManagedSeed{}
+	if err := gardenClient.Client().Get(r.ctx, request.NamespacedName, managedSeed); err != nil {
+		if apierrors.IsNotFound(err) {
+			r.logger.Debugf("Skipping ManagedSeed %s because it has been deleted", request.NamespacedName)
+			return reconcile.Result{}, nil
+		}
+		r.logger.Errorf("Could not get ManagedSeed %s from store: %v", request.NamespacedName, err)
+		return reconcile.Result{}, err
+	}
+
+	if managedSeed.Spec.Shoot == nil || managedSeed.Spec.Shoot.Name == "" {
+		r.logger.Errorf("Skipping ManagedSeed %s because it does not specify a shoot", request.NamespacedName)
+		return reconcile.Result{}, nil
+	}
+
+	if managedSeed.DeletionTimestamp != nil {
+		return r.delete(gardenClient, managedSeed)
+	}
+	return r.createOrUpdate(gardenClient, managedSeed)
+}
+
+func (r *reconciler) createOrUpdate(gardenClient kubernetes.Interface, managedSeed *seedmanagementv1alpha1.ManagedSeed) (reconcile.Result, error) {
+	managedSeedLogger := logger.NewFieldLogger(r.logger, "managedSeed", kutil.ObjectName(managedSeed))
+
+	// Ensure gardener finalizer
+	if err := controllerutils.EnsureFinalizer(r.ctx, gardenClient.DirectClient(), managedSeed, gardencorev1beta1.GardenerName); err != nil {
+		managedSeedLogger.Errorf("Could not ensure gardener finalizer on ManagedSeed: %+v", err)
+		return reconcile.Result{}, err
+	}
+
+	// Get shoot
+	shoot := &gardencorev1beta1.Shoot{}
+	if err := gardenClient.Client().Get(r.ctx, kutil.Key(managedSeed.Namespace, managedSeed.Spec.Shoot.Name), shoot); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Check if the shoot is reconciled
+	if shoot.Generation != shoot.Status.ObservedGeneration || shoot.Status.LastOperation == nil || shoot.Status.LastOperation.State != gardencorev1beta1.LastOperationStateSucceeded {
+		managedSeedLogger.Infof("Waiting for shoot %s to be reconciled before registering it as seed", kutil.ObjectName(shoot))
+
+		// Update status to Pending
+		if updateErr := updateStatusPending(r.ctx, gardenClient.DirectClient(), managedSeed, "Reconciliation pending"); updateErr != nil {
+			managedSeedLogger.Errorf("Could not update ManagedSeed status for pending creation or update: %+v", updateErr)
+			return reconcile.Result{}, updateErr
+		}
+
+		// Return success result with requeue after 10 seconds
+		return reconcile.Result{RequeueAfter: 10 * time.Second}, nil
+	}
+
+	// Get seed client
+	seedClient, err := r.clientMap.GetClient(r.ctx, keys.ForSeedWithName(*shoot.Spec.SeedName))
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("could not get seed client: %w", err)
+	}
+
+	// Get shoot client
+	shootClient, err := r.clientMap.GetClient(r.ctx, keys.ForShoot(shoot))
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("could not get shoot client: %w", err)
+	}
+
+	// Create actuator
+	a := newActuator(gardenClient, seedClient, shootClient, r.config, r.imageVector, r.logger)
```

Comment:
I didn't completely get the motivation behind separating reconciler and actuator here.
I would propose, to either move all functions of the actuator to the reconciler or create only one actuator instance and pass all the required stuff.
I think, creating one actuator per reconciliation is not the best idea and can be easily avoided here.

Reply:
I borrowed the "pattern" from e.g. https://github.com/gardener/gardener/tree/master/pkg/gardenlet/controller/backupbucket (there are a few controllers like this). My idea was simply "separation of concerns" - even though we create one actuator per reconciliation these two types have different responsibilities, so separating them makes the code more modular, and therefore easier to read and maintain. This is why I don't really like "move all functions of the actuator to the reconciler". 

Regarding "create only one actuator instance and pass all the required stuff." I wanted to do it but then I would have to pass the `seedClient` and the `shootClient` in `Reconcile` and `Delete` (and from then on to all methods they invoke, which is quite a number) since they are different with each reconciliation, only the `gardenClient` is always the same. I don't like this either, I see no benefit that would justify the added verbosity.

What is the actual issue you have with "creating one actuator per reconciliation"? I don't think it has any performance or other impact compared to "move all functions of the actuator to the reconciler", but it's cleaner and easier to understand and maintain IMO, plus it's already how other gardenlet controllers are built.

Reply:
Ok, you convinced me :)
I agree, that the separation makes sense and helps with the readability.
Probably, there is indeed not much of a performance penalty for constructing a new `actuator` everytime.